### PR TITLE
[avmfritz] Incorporated new QuantityType (Units of Measurement)

### PR DESCRIPTION
**Pull request by cweitkamp**
- Incorporated new QuantityType (Units of Measurement)
- Removed `org.hamcrest` dependency from unit tests

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -37,5 +39,6 @@ Import-Package:
  org.openhab.binding.avmfritz.handler,
  org.osgi.framework,
  org.osgi.service.component,
- org.slf4j
+ org.slf4j,
+ tec.uom.se
```
Comment (wborn) [inline addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: Is this import necessary? I don't get any hits on the when I search for imports of that package.

Comment (kaikreuzer) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: That package actually must not be imported by bindings.

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: If I do so I get an error while compiling. What has to be done to resolve that? Is there a relation to the `WATT_HOUR` unit used in https://github.com/openhab/openhab2-addons/pull/3477/files#diff-bd72c5c4d5010dd42a9b6940a70ad50bR220? Or did I something wrong? If I remove that unit. It compiles w/o any errors.
```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.0.0:compile (default-compile) on project org.openhab.binding.avmfritz: Compilation failure: Compilation failure:
[ERROR] ~/openhab2-addons/addons/binding/org.openhab.binding.avmfritz/src/main/java/org/openhab/binding/avmfritz/internal/ahamodel/PowerMeterModel.java:
[ERROR] /**
[ERROR] ^
[ERROR] The type tec.uom.se.AbstractSystemOfUnits cannot be resolved. It is indirectly referenced from required .class files
```

Comment (kaikreuzer) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: @htreu That sounds like a bug in ESH, doesn't it?

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: One more thing: It seems to be a problem with the static import of `org.eclipse.smarthome.core.library.unit.SmartHomeUnits.*`.

Comment (htreu) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: investigating. stay tuned.

Comment (htreu) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: Thats actually a bug: https://github.com/eclipse/smarthome/issues/5413

Comment (htreu) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: @cweitkamp the workaround is to **not** statically import `SmartHomeUnits.*` or even single fields like `SmartHomeUnits.WATT_HOUR` but use the simple class import and the qualified static access in the code.

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.avmfritz/META-INF/MANIFEST.MF]: @htreu Thank you. I have changed it for now.

**Comment by cweitkamp** [comment]
No, it is not. You are right. There is an APIBreaking change. The unit of the energy channel has been changed from kWh to Wh.

**Comment by martinvw** [comment]
We document them at https://github.com/openhab/openhab-distro/wiki/%5BDRAFT%5D-openHAB-2.3-Release-Notes

**Comment by wborn** [comment]
Can you also update the release notes now the PR has been merged @cweitkamp?

**Comment by cweitkamp** [comment]
@wborn Done. I added a line to the release notes. Thank you too. I looked at your code in the Nest binding to pimp my work. ;-)

**Comment by kaikreuzer** [comment]
@cweitkamp May I ask for the reason for that breaking change? Imho, kWh is more commonly used than Wh, so from a user perspective this feels like a step backwards to me.
And if there isn't any strong reason for breaking compatibility, we should always choose not to do so.

**Comment by cweitkamp** [comment]
@kaikreuzer Of course. The AHA interface is capable of delivering the data in Wh. In the past we did not have the powerful features of the UoM. Therefore it was the best idea to return the most common and convenient unit for the user - which of course is kWh. My thoughts on this change: Now with the introduction of the UoM we should try to get the best out of it. We can deliver the data in Wh but the user itself can choose which conversion / unit he wants to see in his environment. As a small side note: I didn't touch the unit of this Item in the README.md but leave it in kWh on purpose. If someone follows the example neither the current setup will be affected nor the persistence of the value.

I hope this explanation is sufficient. If not it will be no big deal to revert this regression.

**Comment by htreu** [comment]
If you want the framework to deal with the conversion you could let the thingHandler update the state in the raw unit (`Wh` here) but declare the default state description as `%.3f kWh` just as ist was before [here](https://github.com/openhab/openhab2-addons/pull/3477/files#diff-0dc37b5ac5359232c08fefbf3dab5997R53).
This way any item bound to the channel will get updates in kWh but the framework will do the conversion. If an item overwrites the state description to be `%.3f Wh` it will get the value w/o conversion.

**Comment by kaikreuzer** [comment]
@htreu's suggestion would mean that we do not break backward compatibility - I'd vote for it :-)

**Comment by cweitkamp** [comment]
@htreu @kaikreuzer Thank you very much. That is exactly what I intended to do but I somehow missed that point. Follow-up PR #3490.
I deleted my comment about it in the release notes: https://github.com/openhab/openhab-distro/wiki/%5BDRAFT%5D-openHAB-2.3-Release-Notes.

**Comment by cweitkamp** [comment]
For discussion: @kaikreuzer @htreu 

See also: https://community.openhab.org/t/chart-problem-after-updating-to-openhab-2-3/46376/

After a hint from the community I have reviewed our discussion on the `energy` channel. At the time we will change the source unit of a channel it is imho an APIBreaking change. Why? I will try to explain you: We can manipulate and keep backward compatibility for the UIs by declaring a fixed state description for the item. But we cannot handle persistence in this way. The values stored by the persistence are based on the default unit selected in the users environment. This can lead to issues in e.g. charts. Do you think we can implement a fix for that?

Another example: Lets assume we have a `Number:Speed` item which we want to display in the default speed unit (`m/s`) and in a different speed unit (`km/h`). For both units we want to have a chart in our sitemap. Which currently seems not to be possible. What about a configuration for the unit in charts?

```csv
Number:Speed testNumberSpeed
```

```php
sitemap demo label="My home automation" {
    Frame label="Speed" {
        Text item=testNumberSpeed label="Windgeschwindigkeit [%.1f %unit%]" icon="wind"
        Text item=testNumberSpeed label="Windgeschwindigkeit [%.1f km/h]" icon="wind"
        Chart item=testNumberSpeed label="Windgeschwindigkeit [%.1f %unit%]" period=h legend=true
        Chart item=testNumberSpeed label="Windgeschwindigkeit [%.1f km/h]" period=h legend=true
    }
}
```

bildschirmfoto von 2018-06-15 14-34-38

**Comment by htreu** [comment]
> The values stored by the persistence are based on the default unit selected in the users environment.

Is this correct? I thought the current state of an item is persisted. If a `Number:Speed mySpeed "[%.1f km/h]"` item persists a value of `100 km/h` and the persistence service is not capable of storing the unit (like all of them right now), on restore the item will receive a decimal value of `100` and due to the fixed state description again restore it to `100 km/h`.

> What about a configuration for the unit in charts?

This is currently not supported, but support can be added just like for other widgets (Dimmer, Setpoint, ...). The target unit would again be parsed from the widgets state description. @cweitkamp do you think you can come up with a contribution for that?

**Comment by cweitkamp** [comment]
> > The values stored by the persistence are based on the default unit selected in the users environment.

> Is this correct? I thought the current state of an item is persisted.

No, it is not. Your explanation is the correct way. Thanks for clarification. That is the missing piece in the puzzle. My examples are based on the label of a sitemap element and if we assign an unit there it will of course not taken into account for persistence.

> come up with a contribution for that?

I will put it onto my to-do list.

**Pull request merged**